src/txrequest.cpp

```diff
@@ -0,0 +1,593 @@
+            return false;
+        }
+
+        // Mark the announcement COMPLETED, and select the next best announcement (the first CANDIDATE_READY) if
+        // needed.
+        ChangeAndReselect(it, State::COMPLETED);
+
+        return true;
+    }
+
+    //! Make the data structure consistent with a given point in time:
+    //! - REQUESTED annoucements with expiry <= now are turned into COMPLETED.
+    //! - CANDIDATE_DELAYED announcements with reqtime <= now are turned into CANDIDATE_{READY,BEST}.
+    //! - CANDIDATE_{READY,BEST} announcements with reqtime > now are turned into CANDIDATE_DELAYED.
+    void SetTimePoint(std::chrono::microseconds now)
+    {
+        // Iterate over all CANDIDATE_DELAYED and REQUESTED from old to new, as long as they're in the past,
+        // and convert them to CANDIDATE_READY and COMPLETED respectively.
+        while (!m_index.empty()) {
+            auto it = m_index.get<ByTime>().begin();
+            if (it->m_state == State::CANDIDATE_DELAYED && it->m_time <= now) {
+                PromoteCandidateReady(m_index.project<ByTxHash>(it));
+            } else if (it->m_state == State::REQUESTED && it->m_time <= now) {
+                MakeCompleted(m_index.project<ByTxHash>(it));
+            } else {
+                break;
+            }
+        }
+
+        while (!m_index.empty()) {
+            // If time went backwards, we may need to demote CANDIDATE_BEST and CANDIDATE_READY announcements back
+            // to CANDIDATE_DELAYED. This is an unusual edge case, and unlikely to matter in production. However,
+            // it makes it much easier to specify and test TxRequestTracker::Impl's behaviour.
+            auto it = std::prev(m_index.get<ByTime>().end());
+            if (it->IsSelectable() && it->m_time > now) {
+                ChangeAndReselect(m_index.project<ByTxHash>(it), State::CANDIDATE_DELAYED);
+            } else {
+                break;
+            }
+        }
+    }
+
+public:
+    Impl(bool deterministic) :
+        m_computer(deterministic),
+        // Explicitly initialize m_index as we need to pass a reference to m_computer to ByTxHashViewExtractor.
+        m_index(boost::make_tuple(
+            boost::make_tuple(ByPeerViewExtractor(), std::less<ByPeerView>()),
+            boost::make_tuple(ByTxHashViewExtractor(m_computer), std::less<ByTxHashView>()),
+            boost::make_tuple(ByTimeViewExtractor(), std::less<ByTimeView>())
+        )) {}
+
+    // Disable copying and assigning (a default copy won't work due the stateful ByTxHashViewExtractor).
+    Impl(const Impl&) = delete;
+    Impl& operator=(const Impl&) = delete;
+
+    void DisconnectedPeer(NodeId peer)
+    {
+        auto& index = m_index.get<ByPeer>();
+        auto it = index.lower_bound(ByPeerView{peer, false, uint256::ZERO});
+        while (it != index.end() && it->m_peer == peer) {
+            // Check what to continue with after this iteration. Note that 'it' may change position, and
```

Comment:
Is this safe?
I guess as long as the change of position is moderate (happening within a peer). And perhaps can only go backward? I'm wondering if we may stuck in an endless loop here somehow...

Reply:
Yes, it's safe. The key insight is that no other Announcement objects from the same peer are ever affected by this call (std::next(it) may be affected, but only if it's from a different peer). I've elaborated the comments to hopefully clarify this.

Reply:
Thanks!